src/persist-client/src/internal/state.rs

```diff
@@ -1660,7 +1660,14 @@ where
             return Break(NoOpStateTransition(Since(Antichain::new())));
         }
 
-        let reader_state = self.leased_reader(reader_id);
+        // The only way to have a missing reader in state is if it's been expired... and in that
+        // case, we behave the same as though that reader had been downgraded to the empty antichain.
+        let Some(reader_state) = self.leased_reader(reader_id) else {
+            tracing::warn!(
+                "Leased reader {reader_id} was expired due to inactivity. Did the machine go to sleep?",
+            );
+            return Break(NoOpStateTransition(Since(Antichain::new())));
```

Comment:
While this function does return the new Since, it looks like our caller doesn't check how the returned Since compares to the `new_since` it provided ([code](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/read.rs#L608-L634))?

Thinking out loud a bit here... I imagine theoretically this is okay, but this is what might result in panics elsewhere when someone can no longer read with their handle because the `since` has been advanced to the empty Antichain? Like you call out in the PR description I think this will lead to a net win in stability though, so it works for me!

Reply:
Yeah, that all sounds right!

A slightly more invasive version would be to have an explicit "expired" state, and return a targeted error when folks try and snapshot on an expired handle. I think it mostly comes to the same thing, but may be more explicit... preference?

Reply:
An explicit expired state sounds nice! Definitely don't think it's a blocker though, happy to make that change at a later date